[FEATURE] 토큰 복호화 로직 작성

JSY8869 opened the pull request:
## 제목
Jwt 토큰 복호화 로직 작성

## 작업 내용
1. Jwt token decoder 로직 작성

access 토큰의 경우 prefix를 분리한 후 디코딩 하여 claims의 정보를 통해 유저 정보를 반환하는 방식으로 구현하였습니다.


2. 코드 리팩터링

개발 중 불필요한 코드를 발견하여 제거하였습니다. 
(refresh token creator에서 token의 id를 넣는 부분에서 subject에 user의 id를 넣어주기 때문에 토큰 자체의 id는 필요가 없다고 판단)

3. user service 로직 추가 및 exception 추가

userRepository에 구현되어 있는 findByIdentifier 메서드를 서비스에서 사용 가능하도록 구현하여 토큰 복호화 후 identifier를 이용하여 유저의 정보를 가져올 수 있도록 하였습니다.
이 과정에서 유저를 찾지 못하였을 경우 인가 중 발생한 오류이기 때문에 AuthorizationServiceException 을 상속 받은 UserNotFoundException을 던지도록 구현하였습니다.(이 부분은 어떻게 하는 게 좋을 지 조언 부탁드립니다.)

JSY8869 committed:
feat: Add jwt token decoder

1. TokenDecoder, JwtTokenDecoder 인터페이스 설계
Token 디코딩 관련 로직을 인터페이스를 통해 추상화 하고, 이름에 맞게 책임과 역할을 분리하여 구현하였습니다.

2. Memory 유저에 대한 jwt 토큰 디코더 구현
access 토큰의 경우 prefix가 있기 때문에 이를 분리한 후 토큰을 디코딩 하여 claims의 정보를 통해 유저 정보를 반환하는 방식으로 구현하였습니다.

related to #22

JSY8869 committed:
feat: Add memory user find by identifier in service

1. userService에 findByIdentifier 추가
유저를 찾지 못했을 경우 인가 과정에서 발생하는 에러이기 때문에 AuthorizationServiceException 을 상속 받은 UserNotFoundException 을 throw 하도록 설계하였습니다.

related to #22

JSY8869 committed:
feat: Add decoding function in service

1. token 서비스에서 decoder를 통해 토큰을 디코딩 하도록 설계

related to #22

JSY8869 committed:
feat: Add configuration about token decoder

1. token 디코더에 대한 configuration 추가

related to #22

JSY8869 committed:
refact: Remove unnecessary code

1. 불필요한 코드 제거

related to #22

CokeLee777 commented on src/main/java/io/github/cokelee777/springsecurityjwtauth/security/token/decoder/MemoryJwtTokenDecoder.java:
```diff
@@ -0,0 +1,46 @@
+
+    private final MemoryUserService memoryUserService;
+
+    private JwtParser jwtAccessTokenParser = Jwts.parserBuilder()
```
여기에 final 키워드를 붙이면 좋을 듯 하네요

CokeLee777 commented on src/main/java/io/github/cokelee777/springsecurityjwtauth/security/token/decoder/MemoryJwtTokenDecoder.java:
```diff
@@ -0,0 +1,46 @@
+
+        MemoryUser memoryUser = memoryUserService.findByIdentifier((String) body.get("identifier"));
+
+        return new JwtMemoryUserDetails(memoryUser);
```
클래스 명이 디코더니깐 여기서는 반환값이 Claims가 되고 서비스에서 실제 UserDetails를 반환하는게 좋을듯 한데 어떻게 생각하시나요?

JSY8869 replied:
그렇게 되면 token 서비스에서 user의 정보를 찾는 코드가 들어가야 해서 token service의 책임의 범위가 너무 넓어지기 때문에 claim을 userdetail로 바꾸어주는 클래스를 설계하여 연결하는 방식으로 구현 해보겠습니다.

CokeLee777 commented on src/main/java/io/github/cokelee777/springsecurityjwtauth/security/token/decoder/MemoryJwtTokenDecoder.java:
```diff
@@ -0,0 +1,46 @@
+        // Token prefix 와 실제 토큰 값을 분리
+        String rawAccessToken = accessToken.split(ACCESS_TOKEN_PREFIX)[1];
+
+        Claims body = jwtAccessTokenParser.parseClaimsJws(rawAccessToken).getBody();
```
이 구문에서 예외가 많이나는데, try catch 문으로 받아서 모든 예외에 해당하는 예외 메시지를 던져주면 좋을 것 같네요

JSY8869 replied:
이 부분에서 발생할 예외까지 검증하고 보내주는 건가 했는데 아니었군요ㅎㅎ
예외는 조작된 토큰이라는 내용을 포함 하도록 수정 하겠습니다.

CokeLee777 commented on src/main/java/io/github/cokelee777/springsecurityjwtauth/exception/UserNotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+
+import org.springframework.security.access.AuthorizationServiceException;
+
+public class UserNotFoundException extends AuthorizationServiceException {
```
AuthorizationServiceException 은 예외처리 할 때 한번 더 검증해야 할 듯 합니다. AuthorizationServiceException이 AccessDeniedException을 상속받고 있기 때문에 나중에 예외처리할 때 한번 더 보도록 하겠습니다.

JSY8869 replied:
일단 테스트를 위해 IllegalArgumentException 을 상속 받는 걸로 처리한 후 주석으로 TODO를 달아 놓도록 하겠습니다.

CokeLee777 reviewed (commented):
전반적으로 꼼꼼하게 하지 않은 것 같아서 리뷰보고 커맨트 달아주세요~

JSY8869 committed:
refactor: Revise code based on review

1. token decoder의 역할 분담
정확한 역할 분담을 위해 token decoder에서 decoding만 담당 하고 decoding한 claim 을 이용하여 유저의 데이터를 추출하는 것은 extractor에서 담당하도록 변경

2. 예외 처리 추가
token extractor와 token decoder에서 발생할 수 있는 예외에 대해 처리

3. tokenService 의존성 변경
service의 의존성을 줄이기 위해 tokenService에서 여러가지 도메인을 의존하는 것이 아닌 extractor와 Provider만 의존하도록 변경

4. CustomExpiredJwtException 추가
ExpiredJwtException의 경우 예외를 던질 때 header, claim과 같은 정보를 넣어 주어야 하기 때문에 JwtException 을 상속 받는 CustomExpiredJwtException 을 만들어서 만료된 토큰에 대한 예외를 처리함

5. UserNotFoundException 임시 수정
후에 예외 처리 구현 시 확인 후 수정이 필요함

JSY8869 commented:
화상 회의를 통해 나온 의견을 토대로 코드를 수정하였습니다.
리뷰 부탁드립니다!

CokeLee777 commented on src/main/java/io/github/cokelee777/springsecurityjwtauth/security/token/decoder/MemoryJwtTokenDecoder.java:
```diff
@@ -1,46 +1,58 @@
+        } catch (MalformedJwtException e) {
+            throw new MalformedJwtException("조작된 토큰입니다.");
+        } catch (SignatureException e) {
+            throw new MalformedJwtException("토큰 서명 확인에 실패 하였습니다.");
```
이 부분 예외를 잘못 던진것 같네요 수정바랍니다~

CokeLee777 commented on src/main/java/io/github/cokelee777/springsecurityjwtauth/exception/CustomExpiredJwtException.java:
```diff
@@ -0,0 +1,13 @@
+
+import io.jsonwebtoken.JwtException;
+
+public class CustomExpiredJwtException extends JwtException {
```
이 커스텀 예외 클래스를 굳이 만들어야 하나 싶어요. 보니깐 만료된 토큰 예외에 쓰이던데 그냥 `ExpiredJwtException` 과 같이 원래 있던걸 던지면 되지 않을까 싶습니다

JSY8869 replied:
catch에서 잡힌 ExpiredJwtException 에서 값을 꺼내서 message만 수정하면 되는데 잘못 생각했네요... 수정 하겠습니다!
참고로 이 ExpiredJwtException엔 header와 claims가 무조건 들어가야 합니다.

CokeLee777 replied:
확인했습니다~

CokeLee777 commented on src/main/java/io/github/cokelee777/springsecurityjwtauth/security/token/decoder/MemoryJwtTokenDecoder.java:
```diff
@@ -1,46 +1,58 @@
+        } catch (MalformedJwtException e) {
+            throw new MalformedJwtException("조작된 토큰입니다.");
+        } catch (SignatureException e) {
+            throw new MalformedJwtException("토큰 서명 확인에 실패 하였습니다.");
```
이 부분도 동일하게 수정해야할듯 합니다

CokeLee777 commented on src/main/java/io/github/cokelee777/springsecurityjwtauth/security/token/decoder/MemoryJwtTokenDecoder.java:
```diff
@@ -1,46 +1,58 @@
+            // Token prefix 와 실제 토큰 값을 분리
+            String rawAccessToken = accessToken.split(ACCESS_TOKEN_PREFIX)[1];
+            body = jwtAccessTokenParser.parseClaimsJws(rawAccessToken).getBody();
+        } catch (UnsupportedJwtException | NullPointerException e) {
```
NullPointerException이 일어날 수 있는 상황이 없는것 같은데, 왜 예외처리를 했는지 궁금합니다

JSY8869 replied:
`accessToken.split(ACCESS_TOKEN_PREFIX)[1]` 에서 prefix가 없어 split 후 [1]에 접근할 수 없을 경우를 생각해서
`ArrayIndexOutOfBoundsException `을 넣는 다는게 잘못 넣었네요ㅎㅎ 수정 하겠습니다.

CokeLee777 commented on src/main/java/io/github/cokelee777/springsecurityjwtauth/security/token/extractor/MemoryJwtTokenExtractor.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    private void verify(Claims claims, MemoryUser user) {
+        if (user.getId() != claims.getId()
```
이 조건문 전체 다 잘못된 것 같네요. 메모리 유저의 ID는 String 값이므로 equals를 사용해야할 것같고, 나머지도 claims.get() 메서드는 Object 값을 꺼내기 때문에 지금 보면 주소값을 서로 비교하고 있어요. 그리고 claims에 Id값을 넣어주지 않고 subject에 넣은걸로 알고있는데, 더 꼼꼼하게 해야할듯 싶네요

JSY8869 replied:
claims.get() 메서드가 반환 타입을 지정해줘야 하는군요 수정 하겠습니다.
subject도 수정 하겠습니다!
요즘 python을 하다보니 헷갈리네요ㅎㅎ

CokeLee777 reviewed (commented):
전체적으로 수정해야할 부분이 아직도 보입니다. 리뷰보고 커맨트 달아주세요~

JSY8869 committed:
refactor: Revise code based on review

1. extractor refactoring
verify 메서드 내에서 string 값 간의 비교를 위해 equals로 변경, claims.get() 메서드에서 반환 타입을 지정하지 않으면 Object 타입을 반환하기 때문에 String으로 지정

2. revise decoder's exceptions
토큰 만료 시 custom exception이 아닌 ExpiredJwtException에 message만 수정하여 던지도록 변경,
SignatureException에 대해 SignatureException 을 던지도록 변경,
토큰의 prefix가 없을 때 발생할 수 있는 ArrayIndexOutOfBoundsException 변경

related to #22

JSY8869 commented:
전체적으로 수정 완료했습니다!
컨디션이 안 좋을 때 개발을 했더니 상태가 안좋네요ㅎㅎ

CokeLee777 reviewed (approved):
좋은것 같습니다~ 머지해주셔도 될 듯 합니다
